File: linked_list/linkedlist_stack.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

// 연결리스트 스택 구조체 선언
typedef struct node
{
    int data;
    struct node *next;
} stack;

stack *top = NULL;

void push(int data)
{
    stack *new_node = (stack *)malloc(sizeof(stack)) //new_node에 메모리 할당

        if (new_node == NULL)
    {
        printf("stack is full");
        exit(1);
    } // error handling

    new_node->data = data; // data를 new_node.data에 넣어줌
    new_node->next = top;  //next주소를 넣어줌
    top = new_node;        //top(stack 맨 위에 있는 노드)에 new_node 추가
};

int pop()
{
    stack *del;
    int value;

    if (top == NULL)
    {
        printf("stack is empty");
        exit(1);
    }

    value = top->data; //연결리스트 맨 위에 있는 노드(top)의 데이터 (아래에서 출력하기 위해)
    del = top;
    top = top->next;
    free(del); //연결리스트 맨 위에 있는 노드(top) 삭제

    return value;
};

void print_stack()
{
    stack *i;
    stack *count = top;

    i = count;

    while (i != NULL)
    {
        if (i->next == NULL)
            printf("%d\n", i->data);
        else
            printf("%d --> ", i->data);

        i = count->next;
        count = count->next;
    }
};

int main()
{
    printf("----PUSH----\n");
    push(1);
    push(2);
    push(3);
    push(4);
    push(5);

    print_stack();

    printf("----POP----\n");
    pop();
    pop();

    print_stack();
}

File: linked_list/linkedlist_stack.py
# [참고] https://www.sanfoundry.com/python-program-implement-stack-using-linked-list

# linked list 구조체 클래스
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None


class Stack:
    def __init__(self):
        self.head = None  # 연결 목록의 첫 번째 요소를 가리킴

    def push(self, data):
        if self.head is None:
            self.head = Node(data)
        else:
            # 연결리스트의 맨 앞의 노드를 추가
            new_node = Node(data)
            new_node.next = self.head
            self.head = new_node

    def pop(self):
        if self.head is None:
            return None
        else:
            popped_data = self.head.data  # 연결리스트 맨 앞에 있는 노드의 데이터를 반환+해당 노드 제거
            self.head = self.head.next
            return popped_data


a_stack = Stack()  # 인스턴스 생성
while True:
    print('push <value>')
    print('pop')
    print('quit')
    do = input('What would you like to do? ').split()

    operation = do[0].strip().lower()
    if operation == 'push':
        a_stack.push(int(do[1]))
    elif operation == 'pop':
        popped = a_stack.pop()
        if popped is None:
            print('Stack is empty.')
        else:
            print('Popped value: ', int(popped))
    elif operation == 'quit':
        break
